test/unit/math/test_ad.hpp

```diff
@@ -1881,11 +1881,15 @@ template <typename F, typename EigMat>
 void expect_ad_matvar(const F& f, const EigMat& x) {
   ad_tolerances tols;
   expect_ad_matvar_v(tols, f, x);
+
+  stan::math::recover_memory();
```

Comment:
I was getting sanitizer errors without these. Something wasn't freeing itself with trace_quad_prod.

I don't feel great about that because it doesn't seem like `recover_memory` is used anywhere else in this code.

Reply:
I'll take a look at this

Reply:
My suspicion is that the rest of `expect_ad` has these hidden somewhere, either by using `gradient` or `hessian` functions, or other things (and it's okay to add them). `expect_ad_matvar` shouldn't need to leave anything on the stack after it's done.

Reply:
I'm cool with just calling recover memory here. I think it's the safe/correct thing to do